### ISSUE TITLE
TERM variable set to "dumb" from "xterm" when s6-overlay initializes

### ISSUE THREAD

**Issue by blitterated** [just-containers/s6-overlay#485]
# Issue

When s6-overlay initializes, the `TERM` environment variable gets set to `dumb`. This does not happen if you override `/init` with a different `--entrypoint`. I've searched the s6-linux-init, s6-overlay, and s6-rc codebases for the terms "TERM" and "dumb" using `ag`, but nothing comes up except references to handling the SIGTERM signal.

The issue causes strange behaviors with `tmux` running in the container. `tmux` won't run as long as `TERM=dumb`. Setting `TERM=xterm` allow tmux to run, but it appears frozen. It's not, but there's a disconnect between the `tmux` client and the tty.

# Dockerfiles

The issue happens in Ubuntu and Arch. I haven't tried other distributions yet.

Ubuntu:

```Dockerfile
FROM ubuntu

RUN apt update
RUN apt --yes upgrade
RUN apt --yes install curl tar xz-utils
RUN apt --yes install tmux

ARG S6_OVERLAY_VERSION=3.1.2.1
RUN (cd /tmp && curl -LO "https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-noarch.tar.xz")
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
RUN (cd /tmp && curl -LO "https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-x86_64.tar.xz")
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz

ENTRYPOINT ["/init"]
```

Arch:

```Dockerfile
FROM archlinux

RUN pacman -Syu --noconfirm
RUN pacman -S --noconfirm curl
RUN pacman -S --noconfirm tmux

ARG S6_OVERLAY_VERSION=3.1.2.1
RUN (cd /tmp && curl -LO "https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-noarch.tar.xz")
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
RUN (cd /tmp && curl -LO "https://github.com/just-containers/s6-overlay/releases/download/v${S6_OVERLAY_VERSION}/s6-overlay-x86_64.tar.xz")
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz

ENTRYPOINT ["/init"]
```

# Steps to reproduce the issue

Build an image

```sh
docker build -t s6term .
```

### First, run the image without starting s6-overlay

Start a container, but override the ENTRYPOINT

```sh
docker run -it --rm --entrypoint /bin/bash s6term
```

Check the `TERM` variable

```sh
set | grep TERM
```

```text
TERM=xterm
```

Run `tmux`

```sh
tmux new -s foo
```

It works fine. Exit out of the container.

### Second, run the image as normal

Start a container with no override to ENTRYPOINT.

```sh
docker run -it --rm s6term /bin/bash
```

Check the `TERM` variable

```sh
set | grep TERM
```

```text
TERM=dumb
```

Run `tmux`

```sh
tmux new -s foo
```

We get an error:

```text
open terminal failed: terminal does not support clear
```

Change the value of `TERM`

```sh
export TERM=xterm
```

Run `tmux`

```sh
tmux new -s foo
```

Now we're in a situation where `tmux` looks frozen and unresponsive. It doesn't seem to respond to any keypresses, but it does still respond to either `ctrl-b d` to detach or `ctrl-d` to end the session. Either will get you back to the container's shell.

# More Info

I've run the above both in Docker Desktop for Mac and Docker Engine on Xubuntu 22.04 with the same results. I've compared logs generated by the `tmux` client and server both with and without s6 being initialized. They were the same across the board exluding timestamps and some UIDS for client processes and key bindings. That tells me that the `tmux` client and server think everything is fine.

It seems that there is probably more going on than just s6-overlay or s6 changing the `TERM` variable, but I don't know where to look next.

Thanks for everything you do on this project, thanks for looking, and let me know if there's any more info I can provide.

**Comment by skarnet** [comment]
- s6-overlay does not set the TERM variable: it purposefully does not inherit any variables from the host. If you see it set to `dumb`, it's probably bash seeing it unset and setting it to `dumb` by default. If you want a different TERM variable, you can set it in your Dockerfile.
- s6-overlay isn't designed to handle interactive sessions. It tries to accommodate interactive containers as best as it can, but there will inevitably be things that don't work perfectly. The reason for this is that the terminal needs to be closed and reopened during the container boot, because s6-overlay's init process does not want to use it as a control terminal - only the CMD may run with a control terminal.
- This also interacts poorly with the ad-hoc terminal transmission performed by the `docker` command. Behind the scenes, `docker run -it` does very ugly things to pretend that the container is running directly in your terminal - while it absolutely isn't. This makes it very difficult to support interactive sessions that work in all possible cases; s6-overlay is already jumping through several hoops in order for `^C` to work properly in most cases.
- tmux is a *third* layer of terminal transmission, inside the container. At this point, it would be a miracle if terminals worked perfectly through the whole chain.

In conclusion, I'm sorry, but I'm not going to spend any time on this. Interactive CMDs are nominally supported because they seem to be popular for some reason (what on earth are you people doing with containers??), but running tmux is really pushing it. s6-overlay definitely was not made for this.

**Comment by blitterated** [comment]
No worries. I'm definitely pushing the use case boundaries here for both s6 and Docker. It's good for learning both. I appreciate you taking a look, and thanks for all the information. You've given me plenty to work with.

**Comment by blitterated** [comment]
Went digging through source from `/init` forward until I found this line in [`s6-linux-init.c`](https://github.com/skarnet/s6-linux-init/blob/master/src/init/s6-linux-init.c#L75)

```c
fd = open("/dev/tty0", O_RDONLY | O_NOCTTY) ;
```

Double checked in a container with `lsof`, and it was read only. Instead of `/dev/tty0/` it was on `/dev/pts/0`. That makes sense for a container.

```text
COMMAND PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
bash     61 root    0r   CHR  136,0      0t0    3 /dev/pts/0
bash     61 root    1u   CHR  136,0      0t0    3 /dev/pts/0
bash     61 root    2u   CHR  136,0      0t0    3 /dev/pts/0
bash     61 root  255u   CHR  136,0      0t0    3 /dev/pts/0
```

So I tried reopening STDIN as r/w, and it worked.

```sh
export TERM=xterm
exec 0<>/dev/pts/0
tmux
```

Thanks again for your help and info.

**Comment by skarnet** [comment]
That line is never executed in containers. In a container, the tty pointing to the console is read [here](https://github.com/skarnet/s6-linux-init/blob/master/src/init/s6-linux-init.c#L209). So if anything, the problem points to `ttyname(1)` not returning `/dev/pts/0` as it should in your case.
